test/tracer_test.rb

```diff
@@ -142,4 +142,135 @@ def test_default_service
     assert_equal('foo-bar', tracer.default_service)
     tracer.default_service = old_service
   end
+
+  def test_active_span
+    tracer = get_test_tracer
+    span = tracer.trace('something')
+    assert_equal(span, tracer.active_span, 'current span is active')
+    assert_equal(false, tracer.active_span.finished?)
+  end
+
+  def test_trace_all_args
+    tracer = get_test_tracer
+    tracer.set_tags('env' => 'test', 'temp' => 'cool')
+
+    yesterday = Time.now.utc - 24 * 60 * 60
+    tracer.trace('op',
+                 service: 'special-service',
+                 resource: 'extra-resource',
+                 span_type: 'my-type',
+                 start_time: yesterday,
+                 tags: { 'tag1' => 'value1', 'tag2' => 'value2' }) do
+    end
+
+    spans = tracer.writer.spans()
+    assert_equal(1, spans.length)
+    span = spans[0]
+    assert_equal('special-service', span.service)
+    assert_equal('extra-resource', span.resource)
+    assert_equal('my-type', span.span_type)
+    assert_equal(yesterday, span.start_time)
+    assert_equal({ 'env' => 'test', 'temp' => 'cool', 'tag1' => 'value1', 'tag2' => 'value2' }, span.meta)
+  end
+
+  def test_start_span_all_args
+    tracer = get_test_tracer
+    tracer.set_tags('env' => 'test', 'temp' => 'cool')
+
+    yesterday = Time.now.utc - 24 * 60 * 60
+    span = tracer.start_span('op',
+                             service: 'special-service',
+                             resource: 'extra-resource',
+                             span_type: 'my-type',
+                             start_time: yesterday,
+                             tags: { 'tag1' => 'value1', 'tag2' => 'value2' })
+    span.finish
+
+    spans = tracer.writer.spans()
+    assert_equal(1, spans.length)
+    span = spans[0]
+    assert_equal('special-service', span.service)
+    assert_equal('extra-resource', span.resource)
+    assert_equal('my-type', span.span_type)
+    assert_equal(yesterday, span.start_time)
+    assert_equal({ 'env' => 'test', 'temp' => 'cool', 'tag1' => 'value1', 'tag2' => 'value2' }, span.meta)
+  end
+
+  def test_start_span_child_of_span
+    tracer = get_test_tracer
+
+    root = tracer.start_span('a')
+    root.finish
+
+    spans = tracer.writer.spans()
+    assert_equal(1, spans.length)
+    a = spans[0]
+
+    tracer.trace('b') do
+      span = tracer.start_span('c', child_of: root)
+      span.finish
+    end
+
+    spans = tracer.writer.spans()
+    assert_equal(2, spans.length)
+    b, c = spans
+
+    refute_equal(a.trace_id, b.trace_id, 'a and b do not belong to the same trace')
+    refute_equal(b.trace_id, c.trace_id, 'b and c do not belong to the same trace')
+    assert_equal(a.trace_id, c.trace_id, 'a and c belong to the same trace')
+    assert_equal(a.span_id, c.parent_id, 'a is the parent of c')
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  def test_start_span_child_of_context
+    # skipping this for Ruby 1.9, as it has some threading issues (segfaults on msgpack)
```

Comment:
OK, this sounds scary but it's really really a corner case I hit when using Ruby 1.9. Most of the code running runs fine, as long as you don't play with the fire and mix threads. What's happening here is that I'm submitting data in the main thread, that was created in another thread, and this causes a segfault with msgpack. And, as a side note, a Ruby version for which support has been dropped in years. So forwarding the problem upstream won't do any good. One alternative would be to have another context provider, that does not uses threads but another carrier, and do this test with it. I would typically rewrite this test when such a context provider is implemented for real.